frame/oracle/src/lib.rs

```diff
@@ -955,56 +947,76 @@ pub mod pallet {
 			});
 		}
 
-		// REVIEW: indexing
-		#[allow(clippy::indexing_slicing)] // to get CI to pass
 		pub fn get_twap(
 			asset_id: T::AssetId,
-			price_weights: Vec<Percent>,
+			twap_window: usize,
 		) -> Result<T::PriceValue, DispatchError> {
-			let precision = Percent::from_percent(100);
 			let historical_prices = Self::price_history(asset_id);
 
-			// add an extra to account for current price not stored in history
-			ensure!(historical_prices.len() + 1 >= price_weights.len(), Error::<T>::DepthTooLarge);
+			ensure!(twap_window <= historical_prices.len() + 1, Error::<T>::DepthTooLarge);
 
-			let sum = price_weights
+			let mut prices: Vec<_> = historical_prices
 				.iter()
-				.fold(Some(Percent::zero()), |a, b| a.and_then(|a| a.checked_add(b)));
-			ensure!(sum == Some(precision), Error::<T>::MustSumTo100);
-
-			let mut price_weights: Vec<T::PriceValue> =
-				price_weights.into_iter().map(|v| (v.deconstruct() as u128).into()).collect();
+				.rev()
+				.enumerate()
+				.filter_map(|(i, price)| if i < twap_window { Some(price) } else { None })
+				.rev()
+				.collect();
+			let current_price =
+				Prices::<T>::try_get(asset_id).map_err(|_| Error::<T>::PriceNotFound)?;
+			prices.push(&current_price);
+
+			let weights: Vec<u128> = prices
+				.windows(2)
+				.map(|w| {
+					w.get(0)
+						.and_then(|a| w.get(1).map(|b| (a, b)))
+						.and_then(|(a, b)| b.block.checked_sub(&a.block))
```

Comment:
nit: more descriptive variable names would be good, i have no idea what `a` and `b` are here